internal/otelutil/otel_tracing.go

```diff
@@ -0,0 +1,117 @@
+	"go.opentelemetry.io/otel/trace/noop"
+)
+
+// telemetry helpers
+
+func AddException(span trace.Span, err error) {
+	span.AddEvent("exception", trace.WithAttributes(
+		attribute.KeyValue{Key: "exception.type", Value: attribute.StringValue("error")},
+		attribute.KeyValue{Key: "exception.message", Value: attribute.StringValue(err.Error())},
+		attribute.KeyValue{Key: "exception.stacktrace", Value: attribute.StringValue("stacktrace")},
+		attribute.KeyValue{Key: "exception.escaped", Value: attribute.BoolValue(false)},
+	))
+}
+
+// addSpanField adds a field to a span, using the appropriate method for the type of the value.
+func AddSpanField(span trace.Span, key string, value interface{}) {
+	span.SetAttributes(Attributes(map[string]interface{}{key: value})...)
+}
+
+// AddSpanFields adds multiple fields to a span, using the appropriate method for the type of each value.
+func AddSpanFields(span trace.Span, fields map[string]interface{}) {
+	span.SetAttributes(Attributes(fields)...)
+}
+
+// Attributes converts a map of fields to a slice of attribute.KeyValue, setting types appropriately.
+func Attributes(fields map[string]interface{}) []attribute.KeyValue {
+	attrs := make([]attribute.KeyValue, 0, len(fields))
+	for k, v := range fields {
+		kv := attribute.KeyValue{Key: attribute.Key(k)}
+		switch val := v.(type) {
+		case string:
+			kv.Value = attribute.StringValue(val)
+		case int:
+			kv.Value = attribute.IntValue(val)
+		case int64:
+			kv.Value = attribute.Int64Value(val)
+		case float64:
+			kv.Value = attribute.Float64Value(val)
+		case bool:
+			kv.Value = attribute.BoolValue(val)
+		default:
+			kv.Value = attribute.StringValue(fmt.Sprintf("%v", val))
+		}
+		attrs = append(attrs, kv)
+	}
+	return attrs
+}
+
+// Starts a span with no extra fields.
+func StartSpan(ctx context.Context, tracer trace.Tracer, name string) (context.Context, trace.Span) {
+	return tracer.Start(ctx, name)
+}
+
+// Starts a span with a single field.
+func StartSpanWith(ctx context.Context, tracer trace.Tracer, name string, field string, value interface{}) (context.Context, trace.Span) {
+	return tracer.Start(ctx, name, trace.WithAttributes(Attributes(map[string]interface{}{field: value})...))
+}
+
+// Starts a span with multiple fields.
+func StartSpanMulti(ctx context.Context, tracer trace.Tracer, name string, fields map[string]interface{}) (context.Context, trace.Span) {
+	return tracer.Start(ctx, name, trace.WithAttributes(Attributes(fields)...))
+}
+
+func SetupTracing(cfg config.OTelTracingConfig, resourceLibrary string, resourceVersion string) (tracer trace.Tracer, shutdown func()) {
+	if cfg.APIKey != "" {
+		var protocol otelconfig.Protocol = otelconfig.ProtocolHTTPProto
+
+		cfg.APIHost = strings.TrimSuffix(cfg.APIHost, "/")
+		apihost := fmt.Sprintf("%s:443", cfg.APIHost)
+
+		sampleRate := cfg.SampleRate
+		if sampleRate < 1 {
+			sampleRate = 1
+		}
+
+		var sampleRatio float64 = 1.0 / float64(sampleRate)
+
+		headers := map[string]string{
```

Comment:
I know this is straight from R3, but I just noticed this is unconditionally adding the API Key headers, which is not what we say in the docs. We should be adding this APIKeyHeader only if APIKey is non-blank. That way we don't put the header if a user is sending the traces to some other OTel endpoint.

Reply:
There's a check for APIKey on line 80 to only set up a tracer if APIKey is not empty